File: csharp/circular-buffer/CircularBuffer.cs
public class CircularBuffer<T>
{
    private readonly Queue<T> _queue;
    public int Capacity {get; }
    public CircularBuffer(int capacity)
    {
        Capacity = capacity;
        _queue = new Queue<T>(capacity);
    }

    public T Read()
    => _queue.Count == 0 ? throw new InvalidOperationException("Buffer is empty.") : _queue.Dequeue();

    public void Write(T value)
    {
        if (_queue.Count == Capacity)
        {
            throw new InvalidOperationException("Buffer is full.");
        }
        _queue.Enqueue(value);
    }

    public void Overwrite(T value)
    {
        if(_queue.Count == Capacity)
        {
            _queue.Dequeue();
        }
        _queue.Enqueue(value);
    }

    public void Clear()
    {
        _queue.Clear();
    }
}

File: typescript/circular-buffer/circular-buffer.ts
export default class CircularBuffer<T> {
  private buffer: T[];
  private capacity: number;

  constructor(capacity: number) {
    this.capacity = capacity;
    this.buffer = [];
  }

  write(value: T): void {
    if (this.buffer.length >= this.capacity) {
      throw new BufferFullError();
    }
    this.buffer.push(value);
  }

  read(): T {
    if (this.buffer.length === 0) {
      throw new BufferEmptyError();
    }
    return this.buffer.shift()!;
  }

  forceWrite(value: T): void {
    if (this.buffer.length >= this.capacity) {
      this.read();
    }
    this.write(value);
  }

  clear(): void {
    this.buffer = [];
  }
}

export class BufferFullError extends Error {
  constructor() {
    super('Buffer is full');
    this.name = 'BufferFullError';
  }
}

export class BufferEmptyError extends Error {
  constructor() {
    super('Buffer is empty');
    this.name = 'BufferEmptyError';
  }
}
